packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts

```diff
@@ -52,4 +78,84 @@ export class FunctionUrlOrigin extends cloudfront.OriginBase {
       originKeepaliveTimeout: this.props.keepaliveTimeout?.toSeconds(),
     };
   }
-}
\ No newline at end of file
+}
+
+/**
+ * An Origin for a Lambda Function URL with OAC.
+ */
+class FunctionUrlOriginWithOAC extends cloudfront.OriginBase {
+  private originAccessControl?: cloudfront.IOriginAccessControl;
+  private functionUrl: lambda.IFunctionUrl;
+  private readonly props: FunctionUrlOriginWithOACProps;
+
+  constructor(lambdaFunctionUrl: lambda.IFunctionUrl, props: FunctionUrlOriginWithOACProps = {}) {
+    const domainName = cdk.Fn.select(2, cdk.Fn.split('/', lambdaFunctionUrl.url));
+    super(domainName, props);
+    this.functionUrl = lambdaFunctionUrl;
+    this.originAccessControl = props?.originAccessControl;
+
+    this.props = props;
+
+    validateSecondsInRangeOrUndefined('readTimeout', 1, 180, props.readTimeout);
+    validateSecondsInRangeOrUndefined('keepaliveTimeout', 1, 180, props.keepaliveTimeout);
+  }
+
+  protected renderCustomOriginConfig(): cloudfront.CfnDistribution.CustomOriginConfigProperty | undefined {
+    return {
+      originSslProtocols: [cloudfront.OriginSslPolicy.TLS_V1_2],
+      originProtocolPolicy: cloudfront.OriginProtocolPolicy.HTTPS_ONLY,
+      originReadTimeout: this.props.readTimeout?.toSeconds(),
+      originKeepaliveTimeout: this.props.keepaliveTimeout?.toSeconds(),
+    };
+  }
+
+  public bind(scope: Construct, options: cloudfront.OriginBindOptions): cloudfront.OriginBindConfig {
+    const originBindConfig = super.bind(scope, options);
+
+    if (!this.originAccessControl) {
+      this.originAccessControl = new cloudfront.FunctionUrlOriginAccessControl(scope, 'FunctionUrlOriginAccessControl');
+    }
+
+    this.addInvokePermission(scope, options);
+
+    if (!this.isCheckAuthType()) {
```

Comment:
Thanks for adding the validation as requested @watany-dev ! Could we rename the method to `validateAuthType` for consistency? We are still getting this PR security reviewed as well, so we may ask to throw an error instead of warning here for best security practices. But I will confirm first so you can hold off on that change for now.

Reply:
fixed. https://github.com/aws/aws-cdk/pull/31339/commits/f95862117efce6ccf185b7f320f294f5005a2802